src/go/configgenerator/filtergen/util_test.go

```diff
@@ -379,3 +421,123 @@ func TestGetAPIKeySystemParametersBySelectorFromOPConfig(t *testing.T) {
+	if err != nil {
+		return nil, fmt.Errorf("failed to marshal source file into any: %v", err)
+	}
+	return content, nil
+}
+
+func setupInvalidSourceFileContent(rawDescriptor []byte) *anypb.Any {
+	content := &anypb.Any{Value: rawDescriptor}
+	return content
+}
+
+func TestGetDescriptorBinFromOPConfig(t *testing.T) {
+	rawDescriptor, err := proto.Marshal(validDescriptors)
+	if err != nil {
+		t.Fatalf("Failed to marshal FileDescriptorSet: %v", err)
+	}
+	content, err := setupSourceFileContent(rawDescriptor)
+	if err != nil {
+		t.Fatalf("Failed to setup source file content: %v", err)
+	}
+
+	invalidRawDescriptor := []byte("invalid_random_descriptor")
+	validContentwithInvalidProto, err := setupSourceFileContent(invalidRawDescriptor)
+	if err != nil {
+		t.Fatalf("Failed to setup source file content: %v", err)
+	}
+
+	invalidContentwithInvalidProto := setupInvalidSourceFileContent(invalidRawDescriptor)
+
+	testdata := []struct {
+		Desc            string
+		ServiceConfigIn *servicepb.Service
+		DescriptorBin   []byte
+		WantError       bool
+	}{
+		{
+			Desc: "Successfully extracted the descriptor bytes from the OP service config.",
+			ServiceConfigIn: &servicepb.Service{
+				Name: "bookstore.endpoints.project123.cloud.goog",
+				Apis: []*apipb.Api{
+					{
+						Name: "Anotherlibrary",
+						Methods: []*apipb.Method{
+							{
+								Name: "ListBooks",
+							},
+						},
+					},
+				},
+				SourceInfo: &servicepb.SourceInfo{
+					SourceFiles: []*anypb.Any{content},
+				},
+			},
+			DescriptorBin: rawDescriptor,
+			WantError:     false,
+		},
+		{
+			Desc: "Failure in extracting descriptor bytes from the OP service config. Valid content with invalid proto.",
+			ServiceConfigIn: &servicepb.Service{
+				Name: "bookstore.endpoints.project123.cloud.goog",
+				Apis: []*apipb.Api{
+					{
+						Name: "Anotherlibrary",
+						Methods: []*apipb.Method{
+							{
+								Name: "ListBooks",
+							},
+						},
+					},
+				},
+				SourceInfo: &servicepb.SourceInfo{
+					SourceFiles: []*anypb.Any{validContentwithInvalidProto},
+				},
+			},
+			WantError: false,
+		},
+		{
+			Desc: "Failure in extracting descriptor bytes from the OP service config. Invalid content with invalid proto.",
+			ServiceConfigIn: &servicepb.Service{
+				Name: "bookstore.endpoints.project123.cloud.goog",
+				Apis: []*apipb.Api{
+					{
+						Name: "Anotherlibrary",
+						Methods: []*apipb.Method{
+							{
+								Name: "ListBooks",
+							},
+						},
+					},
+				},
+				SourceInfo: &servicepb.SourceInfo{
+					SourceFiles: []*anypb.Any{invalidContentwithInvalidProto},
+				},
+			},
+			WantError: true,
+		},
+	}
+	for _, tc := range testdata {
+		t.Run(tc.Desc, func(t *testing.T) {
+			descriptorBin, err := GetDescriptorBinFromOPConfig(tc.ServiceConfigIn)
+			if tc.WantError && err == nil {
+				t.Fatalf("GetDescriptorBinFromOPConfig() got error: %v, want error presence: %v", err, tc.WantError)
+			}
+			if !tc.WantError && err != nil {
```

Comment:
suggest you break this into two separate tests, one for positive flow, one for error flow. The test functions will be a lot simpler.

[go/gotip/episodes/50](https://goto.google.com/gotip/episodes/50)